### PR TITLE
Fix for matching inj modul triangle axiom

### PR DESCRIPTION
**Pull request by traiansf**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs [hunk 1]
```diff
@@ -397,18 +398,18 @@ matchNonVarToPattern tools first second
...
+            -- variables here, and we're converting everything else to
+            -- predicates.
             -- TODO: Make a function for this.
-            (finalPredicate, _proof) =
-                makeAndPredicate
-                    predicate
-                    (substitutionToPredicate substitution)
+            leftVars = pureFreeVariables Proxy first
+            (leftSubst, rightSubst) =
+                span ((`elem` leftVars).fst) substitution
```
Comment (vladciobanu) [inline src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: I think this needs to be `partition` and not `span`.

File: src/main/haskell/kore/src/Kore/Step/Function/Registry.hs [hunk 2]
```diff
@@ -84,3 +84,4 @@ axiomToIdEvaluatorPair
                         )
                 _ -> Nothing
         Right (RewriteAxiomPattern _) -> Nothing
+
```
Comment (vladciobanu) [inline src/main/haskell/kore/src/Kore/Step/Function/Registry.hs]: As far as I know, we don't have blank lines at the end of files.

File: src/main/haskell/kore/src/Kore/Variables/Free.hs [hunk 3]
```diff
@@ -69,6 +69,9 @@ pureMergeVariables (ForallPattern f) =
     Set.insert (forallVariable f) (forallChild f)
 pureMergeVariables p = fold p  -- default rule
 
+{-| Extracts free variables in a pure pattern as a set
+of pure (not 'Unified') variables.
```
Comment (vladciobanu) [inline src/main/haskell/kore/src/Kore/Variables/Free.hs]: Should we have a rule for `{-|` vs `--|` comments? I've seen both used for both short and long comments.

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs [hunk 4]
```diff
@@ -397,18 +398,18 @@ matchNonVarToPattern tools first second
...
+            finalPredicate = PredicateSubstitution.toPredicate
+                PredicateSubstitution { predicate, substitution = rightSubst }
         return . return $ -- MonadCounter m => m (Maybe a)
             PredicateSubstitution
                 { predicate = finalPredicate
-                , substitution = []
+                , substitution = leftSubst
```
Comment (vladciobanu) [inline src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: Should add a test or two for this, especially in light of the `span` error above.

File: src/main/haskell/kore/src/Kore/Variables/Free.hs (hunk 3 above)
Comment (traiansf) [reply src/main/haskell/kore/src/Kore/Variables/Free.hs]: I'm usually following the existing style of the file, and I prefer using `-- |` only for short comments.

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs (hunk 1 above)
Comment (ttuegel) [reply src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: Why is it not `filter`?

**Comment by ttuegel** [comment]
Even with this pull request, I get the following error:
```
kore-exec: Unexpected non-false predicate GenericPredicate (Fix (EqualsPattern (Equals {equalsOperandSort = SortActualSort (SortActual {sortActualName = Id {getId = "SortKItem", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 50})}, sortActualSorts = []}), equalsResultSort = SortActualSort (SortActual {sortActualName = Id {getId = "PREDICATE", idLocation = AstLocationNone}, sortActualSorts = []}), equalsFirst = Fix (ApplicationPattern (Application {applicationSymbolOrAlias = SymbolOrAlias {symbolOrAliasConstructor = Id {getId = "inj", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 31})}, symbolOrAliasParams = [SortActualSort (SortActual {sortActualName = Id {getId = "SortKResult", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 35})}, sortActualSorts = []}),SortActualSort (SortActual {sortActualName = Id {getId = "SortKItem", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 50})}, sortActualSorts = []})]}, applicationChildren = [Fix (VariablePattern (ConfigurationVariable (Variable {variableName = Id {getId = "var_3243", idLocation = AstLocationGeneratedVariable}, variableSort = SortActualSort (SortActual {sortActualName = Id {getId = "SortKResult", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 74})}, sortActualSorts = []})})))]})), equalsSecond = Fix (ApplicationPattern (Application {applicationSymbolOrAlias = SymbolOrAlias {symbolOrAliasConstructor = Id {getId = "inj", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 5418, column = 80})}, symbolOrAliasParams = [SortActualSort (SortActual {sortActualName = Id {getId = "SortAExp", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 5418, column = 84})}, sortActualSorts = []}),SortActualSort (SortActual {sortActualName = Id {getId = "SortKItem", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 5418, column = 96})}, sortActualSorts = []})]}, applicationChildren = [Fix (VariablePattern (AxiomVariable (Variable {variableName = Id {getId = "VarHOLE", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 5418, column = 109})}, variableSort = SortActualSort (SortActual {sortActualName = Id {getId = "SortAExp", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 5418, column = 117})}, sortActualSorts = []})})))]}))})))
when substitutions fromList []
do not cover all variables in left axiom:fromList [Variable {variableName = Id {getId = "VarKResult", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 63})}, variableSort = SortActualSort (SortActual {sortActualName = Id {getId = "SortKResult", idLocation = AstLocationFile (FileLocation {fileName = "/home/ttuegel/kore/src/main/k/working/imp-map/./imp-kompiled/definition.kore", line = 4348, column = 74})}, sortActualSorts = []})}].
```
In other words, it looks like it still cannot correctly apply the triangle axiom in reverse.

**Comment by traiansf** [comment]
@ttuegel in this case I think it shouldn't, as Id is not subsort to KResult. But it may be that instead of error, it should just fail to match and try another rule.  I'll try to remember why exactly did we think it should error there.

**Pull request merged**